### ISSUE TITLE
[new design] Pen for new sidebar

### ISSUE THREAD

**Issue by ciar4n** [joomla/40-backend-template#365]
As requested by @dgt41 I've created a pen for the new sidebar based on the design by @svom @coolcat-creations  https://www.dropbox.com/s/jlgncxdggip22rh/Joomla4-031217-lj-white.pdf?dl=0. 

CSS is rewritten from scratch and is standalone (non BS dependent).  The intent is to use this to create the custom element. 

Sidebar pen for review: https://codepen.io/littlesnippets/pen/4d7312648363235a6b1c9ed9a8df64cf

An a11y review would be appreciated as not at all my forte (A11y Team, @brianteeman).

**Comment by C-Lodder** [comment]
![screeny](https://user-images.githubusercontent.com/2019801/37408767-b36d25d2-2794-11e8-9143-f90cdfc4db73.png)

Slight issue if the secondary nav is heigher than the viewport.

**Comment by brianteeman** [comment]
I am not on the a11y team and you should really ask them to check.

I am not sure if the html is the proposed markup or just for demo but at a minimum this needs fixing               `<span class="fa fa-plus"></span>`

And you should check the shades of blue used here https://webaim.org/resources/contrastchecker/

All the markup for the menu fails a11y in general see https://github.com/joomla/joomla-cms/issues/19763 but its been 16 days since the promised answer

**Comment by dgt41** [comment]
I am trying to figure out how the plus sign/button will work with keyboard. Let me explain:
https://www.w3.org/TR/wai-aria-practices-1.1/examples/menubar/menubar-1/menubar-1.html

the left/right arrow keys are supposed to move you to the previous/next root menu (unless there is a deeper ul). Now we end up with only **one** li tag and 2 a tags inside. I think this should be a li for the name and the plus sign should be a ul/li/a. That way we can keep the expected menu behaviour for keyboard users. (please keep in mind that complying to A11Y is not just making some tab keys follow the menu, it's far more complicate than that, check the link I posted above)

**Comment by brianteeman** [comment]
yes i think you are correct - the plus is another level

**Comment by ciar4n** [comment]
@C-Lodder Kinda tricky one this. It is due to the 2nd level items been absolute positioned taking them out of the flow of the page. Not sure if there is a CSS only solution. 

@brianteeman I've posted to the a11y glip channel

> I think this should be a li for the name and the plus sign should be a ul/li/a.

So I guess any 3rd level menu items on this menu will display as an icon inline to the 2nd level items? Icon set by the menu item class name?

**Comment by brianteeman** [comment]
Please post on a public forum like github and not a private channel. They might actually do something then

**Comment by dgt41** [comment]
> So I guess any 3rd level menu items on this menu will display as an icon inline to the 2nd level items? Icon set by the menu item class name?

not necessarily, I mean the plus button should still be the a tag, so it's more about playing with classes and some js to get it in the same level. But anyhow I'm neither an expert of A11Y, all I wanted to point out is that menus are extremely complicated in the keyboard behaviour (if we really mean to be accessible and not do what bootstrap is doing with the tabs 😂)

**Comment by dgt41** [comment]
@brianteeman there is an issue in their repo: https://github.com/joomla/accessibility/issues/35 so I guess the valuable feedback will be posted there

**Comment by ciar4n** [comment]
So I guess the first concern is what markup to use for the child menu items that have a secondary link (plus icon).  Currently we have...

```
<li class="mainnav-child-li" role="menuitem">
  <a class="mainnav-child-item no-dropdown active" href="#" target="">Users</a>
  <a href="#" class="mainnav-child-item-secondary">
    <span class="fa fa-plus"></span>
  </a>
</li>
```

It is suggested this should be...

```
<li class="mainnav-child-li" role="menuitem">
  <a class="mainnav-child-item no-dropdown active" href="#" target="">Users</a>
  <ul>
    <li>
      <a href="#" class="mainnav-child-item-secondary">
        <span class="fa fa-plus"></span>
      </a>
    </li>
  </ul>
</li>
```

This is very much an a11y question so we can wait a while for some recommendations from the accessibility team.

From a CSS perspective both options can be made work.

**Comment by brianteeman** [comment]
I thought some more about should the plus be another level. I was thinking it should be because my head was still in j3 mode. In reality "articles" and "new article" are the same level. they are just displayed differently

**Comment by zwiastunsw** [comment]
Not necessarily. "Articles" is a link to the article list. "New article" (+) is a link to a subpage.  In the hierarchy, it is a link to a subpage. But I agree that it is possible to put them on one level.

**Comment by ciar4n** [comment]
Pen has been updated in line with suggestions by @zwiastunsw here https://github.com/joomla/accessibility/issues/35#issuecomment-373691332

https://codepen.io/littlesnippets/pen/4d7312648363235a6b1c9ed9a8df64cf

**Comment by brianteeman** [comment]
Not sure if this is related to the latest changes or not

<img width="313" alt="chrome_2018-03-17_12-58-20" src="https://user-images.githubusercontent.com/1296369/37555660-f753ff14-29e2-11e8-8d03-a8ad2916a0ca.png">

**Comment by zwiastunsw** [comment]
Please delete the attribute aria-hidden="true" from the ul tag. 
This attribute deletes all nested lists from the DOM and Accessibility tree.

**Comment by C-Lodder** [comment]
@zwiastunsw - The `sr-only` class os used for screen readers only. So normally you'd do something like this:

```
<span class="fa fa-plus" aria-hidden="true"></span>
<span class="sr-only>Add item</span>
```

This does not make sense at all:
```
<span class="fa fa-plus" class="sr-only>Add item</span>
```

as it will display the icon **and** text on screen readers only.

**Comment by zwiastunsw** [comment]
What you have written does not really make sense. I did not suggest to use "aria-hidden" here. 
Hide the icons Users, Menu, Component, etc. for screen readers. 
And because blind users don't see the plus icon, show them the label.

**Comment by brianteeman** [comment]
@zwiastunsw thats what you wrote in point 2 above

**Comment by zwiastunsw** [comment]
I apologise, if that is not clear. I have not written in point 2: 
`<span class="fa fa-plus" aria-hidden="true"></span>
<span class="sr-only">Add item</span>`.
I just wrote:  `<span class="fa fa-plus" class="sr-only>Add item</span>`. 
You rightly drew attention to this line in your first comments.

**Comment by brianteeman** [comment]
What it should be is 

`          <span class="fa fa-plus" aria-hidden="true"><span class="sr-only">Add item</span></span>
`
or


`<span class="fa fa-plus" aria-hidden="true" title="Add item"><span class="sr-only">Add item</span></span>`

**Comment by zwiastunsw** [comment]
Oh, I can understand. My mistake.  It should be:
<span class="fa fa-plus"><span  class="sr-only>Add item</span></span>.

**Comment by brianteeman** [comment]
no that doesnt hide the icon. It should be exactly what I wrote

**Comment by zwiastunsw** [comment]
You are right. But you don't have to hide this icon. Screen readers will not see it.

**Comment by brianteeman** [comment]
Thats not globaly true. Many screen readers will see it and fontawesome themselves state that you should hide it

**Comment by zwiastunsw** [comment]
I have to agree with you  :)
See: https://fontawesome.com/how-to-use/accessibility - Semantic Icons (Webfont with CSS)

**Comment by brianteeman** [comment]
I know - I already did all of this for J3 ;)

**Comment by brianteeman** [comment]
Re
```
Please delete the attribute aria-hidden="true" from the all ul tags.
This attribute deletes all nested lists from the DOM and Accessibility tree.
```

Isnt the objective here to hide the list from the dom when the menu item is collapsed. In which case it is correct for the tree to have aria-hidden=true when it is collapsed and set to false when it i expanded

**Comment by zwiastunsw** [comment]
I agree.

**Comment by zwiastunsw** [comment]
I will withdraw point 5 from my suggestions for the time being and I have to think about that. 
Screen reader users should have immediate access to the menu, not to the toggle switch and the icon bar.

**Comment by C-Lodder** [comment]
@brianteeman You've nested the span and added `aria-hidden="true"` on the outer one. Won't that hide both span tags rather than just the icon?

**Comment by brianteeman** [comment]
yes - sorry - need more coffee

**Comment by ciar4n** [comment]
Pen amended as suggested..

1. Please delete the attribute aria-hidden="true" from the all ul tags.
This attribute deletes all nested lists from the DOM and Accessibility tree.

2. Add content to each line,
`<span class="fa fa-plus"></span>`
for example Add item. And assign class .sr-only (from Bootstrap) to the span tag.
`<span class="fa fa-plus" class="sr-only>Add item</span>`

3. https://github.com/joomla/40-backend-template/issues/365#issuecomment-373966940

4. Add aria-expanded="false" to each list in the submenu (to ul tag). **The value of the attribute should change to `"true"' when the submenu receives a focus.** (@dgt41 if you don't mind I'll leave this bit to you.)

https://codepen.io/littlesnippets/pen/4d7312648363235a6b1c9ed9a8df64cf

**Comment by ciar4n** [comment]
https://github.com/joomla/40-backend-template/issues/365#issuecomment-373918263 https://github.com/joomla/40-backend-template/issues/365#issuecomment-373042686

This visual issue should be largely amended

**Comment by zwiastunsw** [comment]
At this point it is very good for me. I have access to all items only from the keyboard. Screen reader announces each item to me (NVDA, ChromeVox).
I suggest to add the title attribute, as Brian suggested.
`<span class="fa fa-plus" aria-hidden="true" title="Add item"></span>`.
Now you need to modify javascript - provide keyboard support (arrows), change the status attributes of the drop-down items.

**Comment by ciar4n** [comment]
Title attribute added as suggested. I believe that finishes the markup.

Thank you @zwiastunsw @brianteeman @dgt41 @C-Lodder for the help on this. Good work 👍

@dgt41 I'll hand this over to you to work your magic

**Comment by dgt41** [comment]
So here we go: https://codepen.io/dgt41/pen/jzLGaG

ATM there are some changes that need to be done in the HTML (sorry, didn't catch those earlier)

- The close button on the sub level menu doesn't work (another change in the HTML)
- Keyboard is almost there (there are couple bugs to iron out)
- The menu is based on the 4 JS classes shared in W3C page above: https://www.w3.org/TR/wai-aria-practices-1.1/examples/menubar/menubar-1/menubar-1.html

Refer to https://www.w3.org/TR/wai-aria-practices-1.1/examples/menubar/menubar-1/menubar-1.html to check the keyboard functionality and post back what is not working

**Comment by ylahav** [comment]
First, nice implementation.
one remark:
when the sub-menu is opened - the left / right arrow have to close the sub-menu. Now it close it and got to prev / next menu item - and if this menu item has a sub-menu it open it - it is confusing, in my opinion.

**Comment by dgt41** [comment]
@ylahav the keyboard arrow keys can be switched. What I don't know is if that breaks accessibility as what is defined in the W3C document as expected behaviour is what we're currently doing. For me (JS implementation) will be a simple change in the key values but I'm not the one to decide on the behaviour...

**Comment by brianteeman** [comment]
100% we should be doing what it says in the documentation and not applying our own ideas
